frontend/apps/reader/modules/readerrolling.lua

```diff
@@ -1025,6 +1076,9 @@ function ReaderRolling:handleEngineCallback(ev, ...)
         -- file or re-rendering)
         self:showEngineProgress(1 - args[1]/100) -- unfill progress
     elseif ev == "OnDocumentReady" or ev == "OnSaveCacheFileEnd" then
+        if ev == "OnDocumentReady" then
+            self.ui.document:cacheFlows()
```

Comment:
This has nothing to do in this crengine callback handler (mostly used for loading/rendering progress).
It should better be done in some :onReaderReady() handler I think.

Reply:
I'll have to test again, but I believe this an the footer's `onChangeViewMode()` were needed to ensure proper updating when changing switching between paged and scroll mode (which can change forced pagebreaks between flows and the total page count). Maybe they're not (or can be moved) after the latest changes.

Reply:
Actually, the purpose of this call is to regenerate the cache every time the document is re-rendered (due to any change, like fonts, tweaks, etc.). If I put it in `:onReaderReady`, it's only called once at the beginning. I tried setting a callback in `CreDocument` itself, but I didn't succeed (it was never called).

Reply:
I understand, so it's the same as for the TOC, we rebuild when we notice a rendering change, via:
ReaderRolling:updatePos() > self.ui:handleEvent(Event:new("UpdateToc"))
Except we don't do it onReaderReady, but only the first time it is accessed/needed.

So, you might either add an onUpdateToc() too (and share that event) - or send another one where this one is send. Or may be no need for another event as you're already in ReaderRolling.
So, onReaderReady + that one in ReaderRolling:updatePos().

(Ideally, no Reader/UI callback/stuff in CreDocument - which is a document class that could/should theoretically be usable in other contexts than our Reader.)

Reply:
I'll do this. I understand not wanting this callback mixed with something that has a different purpose, but I still feel that this is something that should be done in `CreDocument`. Having crengine send a signal to `ReaderRolling` and then `ReaderRolling` send it "back" to `CreDocument` looks like an unnecessary roundabout. Isn't there an easy way for `CreDocument` to react to crengine events (specifically a "done rendering" event)?

And another thing. Is there any guaranteed order in which `onUpdateToc` methods of different classes/instances are processed? If I want to make sure that the footer is updated *after* `ReaderToc:onUpdateToc`, can I just add a `ReaderFooter:onUpdateToc` method or should I create a different event in `ReaderToc:onUpdateToc` and make `ReaderFooter` react to that?

Reply:
Answered and continued at https://github.com/koreader/koreader/pull/6847#issuecomment-723243231